awx/main/scheduler/task_manager.py

```diff
@@ -688,16 +687,26 @@ def task_needs_capacity(self, task, tasks_to_update_job_explanation):
         logger.debug("{} couldn't be scheduled on graph, waiting for next cycle".format(task.log_format))
 
     def reap_jobs_from_orphaned_instances(self):
-        # discover jobs that are in running state but aren't on an execution node
-        # that we know about; this is a fairly rare event, but it can occur if you,
-        # for example, SQL backup an awx install with running jobs and restore it
-        # elsewhere
-        for j in UnifiedJob.objects.filter(
-            status__in=['pending', 'waiting', 'running'],
-        ).exclude(execution_node__in=Instance.objects.exclude(node_type='hop').values_list('hostname', flat=True)):
-            if j.execution_node and not j.is_container_group_task:
-                logger.error(f'{j.execution_node} is not a registered instance; reaping {j.log_format}')
-                reap_job(j, 'failed')
+        # discover jobs that are in running state but have an unregistered controller node, execution node, or container group
+        # that we know about; it can occur if a running OCP node misses heartbeat and gets deleted,
+        # or, SQL backup an awx install with running jobs and restore it elsewhere
+
+        # TODO: after merging other resillency changes, add job_explanation to all these cases
+        for task in self.all_tasks:
+            if task.controller_node:
+                if task.controller_node not in self.instances:
+                    self.all_tasks.remove(task)
+                    reap_job(task, 'failed')
+
+            if task.execution_node:
+                if task.execution_node not in self.instances:
+                    self.all_tasks.remove(task)
+                    reap_job(task, 'failed')
+            elif task.instance_group_id:
+                # for container group jobs
+                if task.instance_group_id not in self.instance_groups.pk_ig_map:
+                    self.all_tasks.remove(task)
+                    reap_job(task, 'failed')
```

Comment:
one more nit pick - in either of these cases, we should probably do `task.cancel()`. Maybe in a separate task? I know, it's a weird idea.

Reply:
I'm getting cold feet about the idea of a container group reaper.

https://github.com/ansible/awx/blob/cc1e349ea8ecfb25cc21127751522a0be6acc284/awx/main/models/unified_jobs.py#L699

The field is SET_NULL, so the event of `elif task.instance_group_id:` without the instance in the in-memory mapping should only happen due to race conditions, and isn't a good metric for reaping. Perhaps we could do this if `task.status == 'running'` and the instance group is _Null_, but that just buys us into the task of requesting canceling, as mentioned.